### PR TITLE
Simplified CONNECT code and added unit test

### PR DESCRIPTION
**Pull request by oxtoacart**
This simplifies some of the code in #4 and adds unit tests.

cc: @myleshorton @joesis

### CODE REVIEW

File: proxy_connect.go [hunk 1]
```diff
@@ -28,49 +29,36 @@ type BufferSource interface {
...
 // bufferSource: specifies a BufferSource, leave nil to use default
 //
+// okWaitsForUpstream: specifies whether or not to wait on dialing upstream
+// before responding OK
+//
 // dial: the function that's used to dial upstream
 func CONNECT(
 	idleTimeout time.Duration,
 	bufferSource BufferSource,
+	okWaitsForUpstream bool,
```
Comment (oxtoacart) [inline proxy_connect.go]: I just made this an option to the CONNECT method. The fact that Lantern happens to use false on the client and true on the server is really lantern-specific, whereas this proxy package is really intended to be more general.

File: proxy_connect.go [hunk 2]
```diff
@@ -93,47 +81,54 @@ func (ic *connectInterceptor) connect(ctx context.Context, w http.ResponseWriter
...
+		}
+	}
 
 	// Note - for CONNECT requests, we use the Host from the request URL, not the
 	// Host header. See discussion here:
 	// https://ask.wireshark.org/questions/22988/http-host-header-with-and-without-port-number
-	if ic.client {
-		if downstream, err = ic.hijack(w); err != nil {
-			return err
+	if upstream, err = ic.dial(ctx, "tcp", req.URL.Host); err != nil {
```
Comment (oxtoacart) [inline proxy_connect.go]: Only one copy of dial code.

File: proxy_connect.go [hunk 3]
```diff
@@ -181,7 +176,12 @@ func (ic *connectInterceptor) respondOK(writer io.Writer, req *http.Request, res
 	return nil
 }
 
-func (ic *connectInterceptor) respondHijacked(writer io.Writer, req *http.Request, statusCode int, respHeaders http.Header, body []byte) error {
```
Comment (oxtoacart) [inline proxy_connect.go]: This can be a standalone function, make it so.

**Comment by myleshorton** [comment]
Ha! To some extent this is the approach I was trying to avoid because I think it's more confusing, but I do like the change to always immediately hijack. If part of the idea overall is to allow external use of this library, though, then I think we should just use the approach of having the code always respond with OK because otherwise it will have issues with chrome for anyone else using it. 

I still think overall that's probably the best approach (unified code that always responds with OK) because without you end up with the server side responding with a Bad Gateway under the following conditions:

1) Bogus domain (very common in our logs)
2) Timeout connecting to domain due to
     a) The domain actually not accepting connections
     b) An overloaded proxy

In case 1, the Lantern client can quickly switch to a new proxy, but because it's a bogus domain, that proxy will also fail and will also erroneously get marked as bad, as will the next and the next. This is already currently causing issues with bad error counts in borda for proxies that are really working fine, and in my anecdotal experience this seems like the cause of most "errors" in proxies when they're not really errors.

In case 2a, the destination timing out will likely continue to timeout or at least take awhile with future proxies, and it's unlikely the combination of those timeouts won't hit some other client-side timeout before ever reaching the destination anyway. The client will also mark the proxy as bad when again there's not actually anything we can do, again polluting our proxy error data.

Case 2b to me is likely fairly rare because it's just as likely the proxy never would have accepted the connection to begin with. Theoretically, though, this is the case where the client could try another proxy, but you're again running into timeout issues where by the time that first proxy has timed out, it's not clear some other client-side timeout wouldn't kick in that would negate any future call to a new proxy. I would also argue this is a case we want to avoid at all costs just ops-wise, as the Lantern user experience will suffer whenever these proxies are live.

Maybe there are other cases I'm not thinking about?

**Comment by myleshorton** [comment]
Been chatting about this on slack and decided to go with it and to consolidate the code if we ever switch the server to respond with a preemptive OK.

**Pull request merged**